libcontainer/specconv/spec_linux.go

```diff
@@ -538,8 +538,10 @@ func checkPropertyName(s string) error {
 	if len(s) < 3 {
 		return errors.New("too short")
 	}
-	// Check ASCII characters rather than Unicode runes.
-	for _, ch := range s {
+	// Check ASCII characters rather than Unicode runes,
+	// so we have to use indexes rather than range.
+	for i := 0; i < len(s); i++ {
+		ch := s[i]
 		if (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') {
```

Comment:
Could also use the `unicode` utilities for this - seems to be even (but marginally) faster 🤔 (but worth verifying on your machine if you see that as well)

```
BenchmarkCheckPropertyName
BenchmarkCheckPropertyName-8          	 1904448	       617.9 ns/op	     320 B/op	      12 allocs/op
BenchmarkCheckPropertyNameUnicode
BenchmarkCheckPropertyNameUnicode-8   	 1977055	       594.7 ns/op	     320 B/op	      12 allocs/op
```

```go
func checkPropertyName(s string) error {
	if len(s) < 3 {
		return fmt.Errorf("too short")
	}
	// Check ASCII characters rather than Unicode runes,
	// so we have to use indexes rather than range.
	for i := 0; i < len(s); i++ {
		ch := s[i]
		if ch > unicode.MaxASCII || !unicode.IsLetter(rune(ch)) {
			return fmt.Errorf("contains non-alphabetic character")
		}
	}
	return nil
}
```

Reply:
Actually, not sure if that makes sense (covering to a `rune()`); guess that would've been more accurate if we kept the `range` loop

Reply:
Not sure how you managed to get faster results, as obviously two calls and a conversion is more expensive than 4 byte checks. This is probably the effect of dynamic power management / CPU frequency scaling. I work around that by doing a few rounds, and checking that numbers are consistent (benchstat checks that, too), and I usually skip the first measurement (of course, the more correct approach would be to disable CPU freq scaling).

Here's what I see on my machine (tested each implementation 3 times, first round discarded, `benchstat` output):
```
name                 old time/op    new time/op    delta
CheckPropertyName-4     122ns ± 0%     152ns ± 1%   ~     (p=0.100 n=3+3)

name                 old alloc/op   new alloc/op   delta
CheckPropertyName-4     64.0B ± 0%     96.0B ± 0%   ~     (p=0.100 n=3+3)

name                 old allocs/op  new allocs/op  delta
CheckPropertyName-4      4.00 ± 0%      6.00 ± 0%   ~     (p=0.100 n=3+3)
```

Here's the diff:

```diff
diff --git a/libcontainer/specconv/spec_linux.go b/libcontainer/specconv/spec_linux.go
index fad7b802..151174e1 100644
--- a/libcontainer/specconv/spec_linux.go
+++ b/libcontainer/specconv/spec_linux.go
@@ -11,6 +11,7 @@ import (
        "strings"
        "sync"
        "time"
+       "unicode"
 
        systemdDbus "github.com/coreos/go-systemd/v22/dbus"
        dbus "github.com/godbus/dbus/v5"
@@ -542,10 +543,9 @@ func checkPropertyName(s string) error {
        // so we have to use indexes rather than range.
        for i := 0; i < len(s); i++ {
                ch := s[i]
-               if (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') {
-                       continue
+               if ch > unicode.MaxASCII || !unicode.IsLetter(rune(ch)) {
+                       return errors.New("contains non-alphabetic character")
                }
-               return errors.New("contains non-alphabetic character")
        }
        return nil
 }
```

In any case, the difference is negligible -- and as I said, this PR is mostly to fix the comment (by making the code in line with what the comment says).

> Actually, not sure if that makes sense (covering to a `rune()`); guess that would've been more accurate if we kept the `range` loop

I'd do the same if I would use unicode functions, but it does make sense in the first place.

Reply:
> This is probably the effect of dynamic power management / CPU frequency scaling.

Yes, perhaps; I ran it a couple of times, and was somewhat scratching my head.

Reply:
I played a bit with the if statement (there's a way to optimize the checks) but it did not show any difference (probably because the compiled optimizes this anyway).

Here is the (unsuccessful) optimization attempt:

```diff
diff --git a/libcontainer/specconv/spec_linux.go b/libcontainer/specconv/spec_linux.go
index fad7b802..fbe1fc19 100644
--- a/libcontainer/specconv/spec_linux.go
+++ b/libcontainer/specconv/spec_linux.go
@@ -541,11 +541,10 @@ func checkPropertyName(s string) error {
        // Check ASCII characters rather than Unicode runes,
        // so we have to use indexes rather than range.
        for i := 0; i < len(s); i++ {
-               ch := s[i]
-               if (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') {
-                       continue
+               c := s[i]
+               if c < 'A' || c > 'z' || c > 'Z' && c < 'a' {
+                       return errors.New("contains non-alphabetic character")
                }
-               return errors.New("contains non-alphabetic character")
        }
        return nil
 }
```

Reply:
LOL; I was playing with that line as well, but (unfortunately) `A-Z` and `a-z` are distinct ranges (numbers are in between I think?), so I don't think there's a very clean way to do it.

I went looking if there's some smart code in stdlib (which is how I came across the unicode thing), and found [some old code (too complicated)](https://pkg.go.dev/golang.org/x/exp/utf8string#String.IsASCII), some [personal ASCII package](https://pkg.go.dev/github.com/galsondor/go-ascii) (looked neat, but too much _just for this_), and some code in the `strings` package, that didn't do it any different https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/strings/strings.go;l=550-559

Reply:
Well, there's a very smart code in strings package, which uses some bit arithmetics, and that _might_ be faster, but again I don't thing this code is worth optimizing more
https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/strings/strings.go;l=794-818;drc=refs%2Ftags%2Fgo1.17.6